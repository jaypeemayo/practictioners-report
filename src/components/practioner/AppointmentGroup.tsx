import * as React from "react";
import { FlexColumn } from "../../common/FlexColumn";
import { FlexRow } from "../../common/FlexRow";
import { Card } from "../../common/Card";
import { IAppointmentGroup } from "./IAppointmentGroup";

export interface IAppointmentGroupProps extends IAppointmentGroup {
}

export const AppointmentGroup = (props: IAppointmentGroupProps) => {
    return (
        <Card title={props.groupKey}>
            <FlexColumn>
                <FlexRow>
                    <div>Cost:</div>
                    <div>{props.cost}</div>
                </FlexRow>
                <FlexRow>
                    <div>Revenue:</div>
                    <div>{props.revenue}</div>
                </FlexRow>
                <FlexRow>
                    <div>Count:</div>
                    <div>{props.count}</div>
                </FlexRow>
            </FlexColumn>
        </Card>
    );
}

