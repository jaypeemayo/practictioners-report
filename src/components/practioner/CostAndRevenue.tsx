import * as React from "react";
import { FlexColumn } from "../../common/FlexColumn";
import { FlexRow } from "../../common/FlexRow";
import { Card } from "../../common/Card";


export interface ICostAndRevenueProps {
    cost: string;
    revenue: string;
    title: string;
}

export const CostAndRevenue = (props: ICostAndRevenueProps) => {
    return (
        <Card title={props.title}>
            <FlexColumn>
                <FlexRow>
                    <div>Cost:</div>
                    <div>{props.cost}</div>
                </FlexRow>
                <FlexRow>
                    <div>Revenue:</div>
                    <div>{props.revenue}</div>
                </FlexRow>
            </FlexColumn>
        </Card>
    );
}

