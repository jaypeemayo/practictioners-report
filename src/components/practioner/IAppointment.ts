export interface IAppointment
{
    clientName: string;
    cost: number;
    date: Date;
    duration: number;
    id: number;
    revenue:number;
    type: string;
}