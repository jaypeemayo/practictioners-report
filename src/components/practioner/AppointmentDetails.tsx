import * as React from "react";
import { IAppointment } from "./IAppointment";

export interface IAppointmentDetailsProps extends IAppointment {
}

export const AppointmentDetails = (props: IAppointmentDetailsProps) =>
    <div className="border border-secondary">
        <h6 className="p-3 mb-2 bg-secondary text-white">Appointments Detail</h6>

        <div className="row">
            <div className="col-6 text-right display">
                <strong>Client Name:</strong>
            </div>
            <div className="col-6">
                {props.clientName}
            </div>

        </div>
        <div className="row">
            <div className="col-6 text-right">
                <strong>Appointment Type:</strong>
            </div>
            <div className="col-6">
                {props.type}
            </div>

        </div>
        <div className="row">
            <div className="col-6 text-right">
                <strong>Duration:</strong>
            </div>
            <div className="col-6">
                {props.duration} min
                </div>
        </div>
    </div>;


