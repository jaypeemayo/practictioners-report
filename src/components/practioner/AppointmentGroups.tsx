import * as React from "react";
import { IAppointment } from "./IAppointment";
import { IAppointmentGroup } from "./IAppointmentGroup";
import { ArrayHelper } from "../../helpers/ArrayHelper";
import { FlexRow } from "../../common/FlexRow";
import { AppointmentGroup } from "./AppointmentGroup";
let moment = require("moment");
if ("default" in moment) {
    moment = moment["default"];
}

export interface IAppointmentGroupsProps {
    appointments: IAppointment[];
    getGroupKey:(item: IAppointment) => any;
}

export class AppointmentGroups extends React.Component<IAppointmentGroupsProps>  {
    getGroupedAppointments = (): IAppointmentGroup[] =>{
        var groups: IAppointmentGroup[] = this.props.appointments.reduce((accumulator: IAppointmentGroup[], item: IAppointment) => {
            let key = this.props.getGroupKey(item);
            let index: number = ArrayHelper.findIndex(accumulator, (g: IAppointmentGroup) => g.groupKey === key);
            if (index >= 0) {
                let group = accumulator[index];
                group = {
                    ...group,
                    cost: group.cost + item.cost,
                    revenue: group.revenue + item.revenue,
                    count: group.count + 1,
                }

                accumulator = ArrayHelper.updateArray(accumulator, index, group);
            }
            else {
                accumulator = [...accumulator,
                { groupKey: key, cost: item.cost, revenue: item.revenue, count: 1 } as IAppointmentGroup]
            }
            return accumulator;

        }, [] as IAppointmentGroup[]);
        return groups;
    }

    render() {
        return (
            <FlexRow>
                {this.getGroupedAppointments().map(group =>
                    <div className={"p-2"}>
                        <AppointmentGroup {...group} />
                    </div>
                )}
            </FlexRow>
        );
    }
}


