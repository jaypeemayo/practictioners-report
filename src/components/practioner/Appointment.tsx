import * as React from "react";
import AppointmentDetails from "./AppointmentDetails";

export interface IAppointmentProps {
    date: string,
    cost: string,
    revenue: string
}

export interface IAppointmentState {
    isAppointmentDetailsShown: boolean;
}

export default class Appointment extends React.Component<IAppointmentProps, IAppointmentState> {
    state: IAppointmentState = { isAppointmentDetailsShown: false };
    onClick = () => {
        this.setState({isAppointmentDetailsShown: !this.state.isAppointmentDetailsShown})
    }
    render() {
        return (
            <>
                <tr onClick={this.onClick}>
                    <th scope="row">{this.props.date}</th>
                    <td>{this.props.cost}</td>
                    <td>{this.props.revenue}</td>

                </tr>
                {this.state.isAppointmentDetailsShown && <tr>
                    <td colSpan={3}>
                        <AppointmentDetails />
                    </td>
                </tr>}
            </>
        );
    }
}
