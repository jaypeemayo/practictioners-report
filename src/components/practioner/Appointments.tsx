import * as React from "react";
import { Table } from "../../common/Table";
import Appointment from "./Appointment";
import { IAppointment } from "./IAppointment";

export interface IAppointmentsProps {
    appointments: IAppointment[];
}

export const Appointments = (props: IAppointmentsProps) =>
    <div className="border border-secondary">
        <h6 className="p-3 mb-2 bg-secondary text-white">Appointments</h6>
        <Table head={<tr>
            <th scope="col">Date</th>
            <th scope="col">Time</th>
            <th scope="col">Cost</th>
            <th scope="col">Revenue</th>
            <th scope="col"></th>
        </tr>}>
            <>
                {props.appointments.map((appointment, i) => <Appointment key={i} {...appointment} />)}
            </>
        </Table>
    </div>;
