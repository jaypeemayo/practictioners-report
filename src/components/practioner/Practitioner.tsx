import * as React from "react";
import { FlexColumn } from "../../common/FlexColumn";
import { Appointments } from "./Appointments";
import { FaChevronUp } from 'react-icons/fa';
import { FaChevronDown } from 'react-icons/fa';
import { IPractitioner } from "./IPractitioner";
import { IAppointment } from "./IAppointment";
import { AppointmentGroups } from "./AppointmentGroups";
let moment = require("moment");
if ("default" in moment) {
    moment = moment["default"];
}
export interface IPractitionerState {
    isAppointmentShown: boolean;
}

export interface IPractitionerProps extends IPractitioner {
}

export default class Practitioner extends React.Component<IPractitionerProps, IPractitionerState> {
    state: IPractitionerState = { startDate: undefined, endDate: undefined, isAppointmentShown: false } as IPractitionerState;
    onClick = () => {
        this.setState({ isAppointmentShown: !this.state.isAppointmentShown });
    }
    renderChevron = () => {
        if (this.state.isAppointmentShown) {
            return <FaChevronUp />;
        }
        else {
            return <FaChevronDown />;
        }
    }

    getGroupKey = (item: IAppointment) => {
        return `${moment(item.date).format("MMM")} ${moment(item.date).format("YYYY")}`;
        //return `Week ${Math.ceil(moment(item.date).date()/7)} of ${moment(item.date).format("MMMM")} ${moment(item.date).format("YYYY")}`;
        //return moment(item.date).format("YYYY");
    }

    render() {
        return (
            <div className="border border-primary">
                <FlexColumn>
                    <h4 className="p-3 mb-2 bg-primary text-white" onClick={this.onClick}>{this.props.name} {this.renderChevron()}</h4>

                    <AppointmentGroups appointments={this.props.appointments} getGroupKey={this.getGroupKey} />

                    {this.state.isAppointmentShown && <div className={"p-2"}>
                        <Appointments appointments={this.props.appointments} />
                    </div>}
                </FlexColumn>
            </div>
        );
    }
}
