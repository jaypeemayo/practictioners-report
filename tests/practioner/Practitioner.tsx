import * as React from 'react';
import * renderer from 'react-test-renderer';
import Practitioner from '../../src/components/practioner/Practitioner';
import { IAppointment } from '../../src/components/practioner/IAppointment';

test('it should handle empty appintments', () => {
    const component = renderer.create(
        <Practitioner name={"DummyName"} appointments={[]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
});

test('should render pratitioners', () => {
    const component = renderer.create(
        <Practitioner name={"DummyName"} appointments={[    {
            clientName: "string",
            cost: 100,
            date: new Date(2000),
            duration: 200,
            id: 1,
            revenue:300,
            type: "Type A"
        } as IAppointment as IAppointment] as IAppointment[]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
});

it('should toggle when practitioners name is clicked', () => {
    const component = renderer.create(
        <Practitioner name={"DummyName"} appointments={[    {
            clientName: "string",
            cost: 100,
            date: new Date(2000),
            duration: 200,
            id: 1,
            revenue:300,
            type: "Type A"
        } as IAppointment as IAppointment] as IAppointment[]}/>,
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      tree.children[0].children[0].props.onClick()
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();
      tree.children[0].children[0].props.onClick()
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();
});
