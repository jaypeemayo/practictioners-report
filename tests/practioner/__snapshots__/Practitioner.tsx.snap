// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should handle empty appintments 1`] = `
<div
  className="border border-primary"
>
  <div
    className="d-flex flex-column "
  >
    <h4
      className="p-3 mb-2 bg-primary text-white"
      onClick={[Function]}
    >
      DummyName
       
      <svg
        attr={
          Object {
            "viewBox": "0 0 448 512",
          }
        }
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </h4>
    <div
      className="d-flex flex-row flex-wrap"
    />
  </div>
</div>
`;

exports[`should render pratitioners 1`] = `
<div
  className="border border-primary"
>
  <div
    className="d-flex flex-column "
  >
    <h4
      className="p-3 mb-2 bg-primary text-white"
      onClick={[Function]}
    >
      DummyName
       
      <svg
        attr={
          Object {
            "viewBox": "0 0 448 512",
          }
        }
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </h4>
    <div
      className="d-flex flex-row flex-wrap"
    >
      <div
        className="p-2"
      >
        <div
          className="border border-secondary"
        >
          <h5
            className="p-3 mb-2 bg-secondary text-white"
          >
            Jan 1970
          </h5>
          <div
            className="p-2"
          >
            <div
              className="d-flex flex-column "
            >
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Cost:
                </div>
                <div>
                  100
                </div>
              </div>
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Revenue:
                </div>
                <div>
                  300
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should toggle when practitioners name is clicked 1`] = `
<div
  className="border border-primary"
>
  <div
    className="d-flex flex-column "
  >
    <h4
      className="p-3 mb-2 bg-primary text-white"
      onClick={[Function]}
    >
      DummyName
       
      <svg
        attr={
          Object {
            "viewBox": "0 0 448 512",
          }
        }
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </h4>
    <div
      className="d-flex flex-row flex-wrap"
    >
      <div
        className="p-2"
      >
        <div
          className="border border-secondary"
        >
          <h5
            className="p-3 mb-2 bg-secondary text-white"
          >
            Jan 1970
          </h5>
          <div
            className="p-2"
          >
            <div
              className="d-flex flex-column "
            >
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Cost:
                </div>
                <div>
                  100
                </div>
              </div>
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Revenue:
                </div>
                <div>
                  300
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should toggle when practitioners name is clicked 2`] = `
<div
  className="border border-primary"
>
  <div
    className="d-flex flex-column "
  >
    <h4
      className="p-3 mb-2 bg-primary text-white"
      onClick={[Function]}
    >
      DummyName
       
      <svg
        attr={
          Object {
            "viewBox": "0 0 448 512",
          }
        }
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
        />
      </svg>
    </h4>
    <div
      className="d-flex flex-row flex-wrap"
    >
      <div
        className="p-2"
      >
        <div
          className="border border-secondary"
        >
          <h5
            className="p-3 mb-2 bg-secondary text-white"
          >
            Jan 1970
          </h5>
          <div
            className="p-2"
          >
            <div
              className="d-flex flex-column "
            >
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Cost:
                </div>
                <div>
                  100
                </div>
              </div>
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Revenue:
                </div>
                <div>
                  300
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="p-2"
    >
      <div
        className="border border-secondary"
      >
        <h6
          className="p-3 mb-2 bg-secondary text-white"
        >
          Appointments
        </h6>
        <table
          className="table thead-light table-hover"
        >
          <thead
            className="thead-light"
          >
            <tr>
              <th
                scope="col"
              >
                Date
              </th>
              <th
                scope="col"
              >
                Time
              </th>
              <th
                scope="col"
              >
                Cost
              </th>
              <th
                scope="col"
              >
                Revenue
              </th>
              <th
                scope="col"
              />
            </tr>
          </thead>
          <tbody>
            <tr
              onClick={[Function]}
            >
              <th
                scope="row"
              >
                01/01/1970
              </th>
              <td>
                10:00 AM
              </td>
              <td>
                100
              </td>
              <td>
                300
              </td>
              <td>
                <svg
                  attr={
                    Object {
                      "viewBox": "0 0 448 512",
                    }
                  }
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 448 512"
                  width="1em"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                  />
                </svg>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`should toggle when practitioners name is clicked 3`] = `
<div
  className="border border-primary"
>
  <div
    className="d-flex flex-column "
  >
    <h4
      className="p-3 mb-2 bg-primary text-white"
      onClick={[Function]}
    >
      DummyName
       
      <svg
        attr={
          Object {
            "viewBox": "0 0 448 512",
          }
        }
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </h4>
    <div
      className="d-flex flex-row flex-wrap"
    >
      <div
        className="p-2"
      >
        <div
          className="border border-secondary"
        >
          <h5
            className="p-3 mb-2 bg-secondary text-white"
          >
            Jan 1970
          </h5>
          <div
            className="p-2"
          >
            <div
              className="d-flex flex-column "
            >
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Cost:
                </div>
                <div>
                  100
                </div>
              </div>
              <div
                className="d-flex flex-row flex-wrap"
              >
                <div>
                  Revenue:
                </div>
                <div>
                  300
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
